src/Giraffe/Routing.fs

```diff
@@ -10,37 +10,97 @@ open FSharp.Control.Tasks.V2.ContextInsensitive
 open Giraffe.FormatExpressions
 
 // ---------------------------
-// Private sub route helper functions
+// Sub Routing Feature
 // ---------------------------
 
-[<RequireQualifiedAccess>]
-module SubRouting =
+[<Literal>]
+let private RouteKey = "giraffe_route"
+
+type ISubRoutingFeature =
+    abstract member CompatibilityMode : GiraffeCompatibilityMode with get
+
+    abstract member GetResolvedPath   : unit   -> string option
+    abstract member SetResolvedPath   : string -> unit
+    abstract member ClearResolvedPath : unit   -> unit
+
+type SubRoutingFeature (compatibilityMode : GiraffeCompatibilityMode) =
+    let mutable resolvedPath = None
 
-    [<Literal>]
-    let private RouteKey = "giraffe_route"
+    interface ISubRoutingFeature with
+        member val CompatibilityMode = compatibilityMode with get
 
-    let getSavedPartialPath (ctx : HttpContext) =
-        if ctx.Items.ContainsKey RouteKey
-        then ctx.Items.Item RouteKey |> string |> strOption
+        member __.GetResolvedPath()             = resolvedPath
+        member __.SetResolvedPath (p : string)  = resolvedPath <- Some p
+        member __.ClearResolvedPath()           = resolvedPath <- None
+
+type HttpContext with
+
+    [<Obsolete("This method has been deprecated. Please use 'GetResolvedPath' instead which is a more memory efficient implementation based on the HttpContext.Features API.")>]
+    member this.GetResolvedPathFromItems() =
+        if this.Items.ContainsKey RouteKey
+        then this.Items.Item RouteKey |> string |> strOption
         else None
 
-    let getNextPartOfPath (ctx : HttpContext) =
-        match getSavedPartialPath ctx with
-        | Some p when ctx.Request.Path.Value.Contains p -> ctx.Request.Path.Value.[p.Length..]
-        | _   -> ctx.Request.Path.Value
+    /// **Description**
+    ///
+    /// Returns a partially (or fully) resolved request path during routing decisions when working with the SubRoutingFeature.
+    ///
+    member this.GetResolvedPath() =
+        let subRoutingFeature = this.Features.Get<ISubRoutingFeature>()
+        match subRoutingFeature.GetResolvedPath() with
+        | Some p -> Some p
+        | None   ->
+            match subRoutingFeature.CompatibilityMode with
```

Comment:
I think this may be the only primitive method that should really exist as an extension on HttpContext, as it's actually used by Giraffe users. The other three not so much (GetNextPartOfPath, SetResolvedPath, ClearResolvedPath) and would be better exposed through a single method into the feature, say `GetSubRouting: unit -> ISubRoutingFeature` .